narwhals/translate.py

```diff
@@ -223,11 +237,192 @@ def get_native_namespace(obj: Any) -> Any:
     return obj.__native_namespace__()
 
 
+def narwhalify(
+    func: Callable[..., Any] | None = None,
+    *,
+    strict: bool = True,
+    eager_only: bool | None = None,
+    series_only: bool | None = None,
+    allow_series: bool | None = None,
+) -> Callable[..., Any]:
+    """
+    Decorate function so it becomes dataframe-agnostic.
+
+    Instead of writing
+
+    ```python
+    import narwhals as nw
+
+    def func(df_any):
+        df = nw.from_native(df_any, strict=False)
+        df = df.group_by('a').agg(nw.col('b').sum())
+        return nw.to_native(df)
+    ```
+
+    you can just write
+
+    ```python
+    import narwhals as nw
+
+    @nw.narwhalify
+    def func(df):
+        return df.group_by('a').agg(nw.col('b').sum())
+    ```
+
+    You can also pass in extra arguments, e.g.
+
+    ```python
+    @nw.narhwalify(eager_only=True)
+    ```
+
+    that will get passed down to `nw.from_native`.
+
+    Arguments:
+        func: Function to wrap in a `from_native`-`to_native` block.
+        strict: Whether to raise if object can't be converted (default) or
+            to just leave it as-is.
+        eager_only: Whether to only allow eager objects.
+        series_only: Whether to only allow series.
+        allow_series: Whether to allow series (default is only dataframe / lazyframe).
+
+    See Also:
+        narwhalify_method: If you want to narwhalify a class method, use that instead.
+    """
+
+    def decorator(func: Callable[..., Any]) -> Callable[..., Any]:
+        @wraps(func)
+        def wrapper(*args: Any, **kwargs: Any) -> Any:
+            if args:
+                df_any = args[0]
+            elif kwargs:
+                params = list(inspect.signature(func).parameters.keys())
+                first_key = params[0]
+                df_any = kwargs[first_key]
+            else:
+                raise TypeError("Expected function which takes at least one argument.")
+            df = from_native(
+                df_any,
+                strict=strict,
+                eager_only=eager_only,
+                series_only=series_only,
+                allow_series=allow_series,
+            )
+            if args:
+                result = func(df, *args[1:], **kwargs)
+            else:
+                kwargs[first_key] = df
+                result = func(**kwargs)
+            return to_native(result, strict=strict)
```

Comment:
Hey @hameerabbasi, thanks for the feedback and reporting this. 

That is totally correct, and even more generally one could return multiple values, dataframes or both, and it is hard to know how we should behave with each one of these.

For now `narwhalify` is intended as a shortcut for most common use case(s) of:

- One dataframe in
- Do operations with it
- Same dataframe out


I believe that for other cases it is better for the users to be _explicit_ in when, how and what to convert.

Reply:
thanks @hameerabbasi for your input!

indeed, as Francesco says, this is just intended as a shortcut for what's probably the most common case, it's not meant to work for every scenario

maybe we should add other shortcuts, like:
- narwhalify_all
- narwhalify_some
- narwhalify_series
- ...

I really want to obsess over user-friendliness with Narwhals

Some people sometimes claim that developer-facing APIs need to be maximally explicit rather than ergonomic, and I completely (but respectfully) disagree with them 😄